pkg/failover/waiter.go

```diff
@@ -0,0 +1,142 @@
+package failover
+
+import (
+	"github.com/spotahome/redis-operator/pkg/log"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+func (r *RedisFailoverKubeClient) waitForPod(name string, namespace string, logger log.Logger) {
+	t := r.clock.NewTicker(loopInterval)
+	defer t.Stop()
+
+	for range t.C {
+		logger.Debug("Waiting for pod to be ready")
+		pod, _ := r.Client.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == "Ready" && condition.Status == v1.ConditionTrue {
+				return
+			}
+		}
+	}
+}
+
+func (r *RedisFailoverKubeClient) waitForService(name string, namespace string, logger log.Logger) {
+	t := r.clock.NewTicker(loopInterval)
+	defer t.Stop()
+
+	for range t.C {
+		logger.Debug("Waiting for service to find bootstrap pod")
+		endpoints, _ := r.Client.CoreV1().Endpoints(namespace).Get(name, metav1.GetOptions{})
+		addresses := 0
+		for _, subset := range endpoints.Subsets {
+			addresses += len(subset.Addresses)
+		}
+		if addresses > 0 {
+			return
+		}
+	}
+}
+
+func (r *RedisFailoverKubeClient) waitForDeployment(name string, namespace string, replicas int32, logger log.Logger) {
+	t := r.clock.NewTicker(loopInterval)
+	defer t.Stop()
+
+	for range t.C {
+		logger.Debug("Waiting for Sentinel deployment to be fully operative")
+		deployment, _ := r.Client.AppsV1beta1().Deployments(namespace).Get(name, metav1.GetOptions{})
+		if deployment.Status.ReadyReplicas == replicas {
+			return
+		}
+	}
+}
+
+func (r *RedisFailoverKubeClient) waitForStatefulset(name string, namespace string, replicas int32, logger log.Logger) {
+	t := r.clock.NewTicker(loopInterval)
+	defer t.Stop()
+
+	for range t.C {
+		logger.Debug("Waiting for Redis statefulset to be fully operative")
+		statefulset, _ := r.Client.AppsV1beta1().StatefulSets(namespace).Get(name, metav1.GetOptions{})
+		if statefulset.Status.ReadyReplicas == replicas {
+			return
+		}
+	}
+}
+
+func (r *RedisFailoverKubeClient) waitForPodDeletion(name string, namespace string, logger log.Logger) {
```

Comment:
All this deletion methods could be reduced to one codebase (no code duplication) and pass an interface taht gets the items.

Reply:
As we've discussed, It's more complicated to do an interface because runtime.Object does not implement the Items function (for the list). We'd have to do a manual object copy and it does not worth it.